logging/writer.go

```diff
@@ -1,29 +1,70 @@
 package logging
 
-import "io"
+import (
+	"bytes"
+	"io"
+	"strings"
+	"sync"
+)
 
 // writer is an implementation of io.Writer that writes to a logger.
 type writer struct {
 	l Logger
 	f func(Logger, string)
+
+	mu  sync.Mutex
+	buf strings.Builder
 }
 
 // NewWriter returns an io.Writer that writes content to the given logger.
 //
-// Each call to Write() produces a separate log message.
+// The string passed to Write() is sliced by end-of-line (EOL) characters and
+// each substring in between the EOL characters is written as a separate log
+// message.
+//
+// If the entire string or its last section does not terminate with an EOL
+// character, it is buffered inside the writer. The writer keeps buffering the
+// string until it detects an EOL character and produces a log message with the
+// buffered string content before the EOL character.
 func NewWriter(l Logger) io.Writer {
-	return &writer{l, LogString}
+	return &writer{l: l, f: LogString}
 }
 
 // NewDebugWriter returns an io.Writer that writes content to the given logger
 // as debug messages.
 //
-// Each call to Write() produces a separate log message.
+// The string passed to Write() is sliced by end-of-line (EOL) characters and
+// each substring in between the EOL characters is written as a separate log
+// message.
+//
+// If the entire string or its last section does not terminate with an EOL
+// character, it is buffered inside the writer. The writer keeps buffering the
+// string until it detects an EOL character and produces a log message with the
+// buffered string content before the EOL character.
 func NewDebugWriter(l Logger) io.Writer {
-	return &writer{l, DebugString}
+	return &writer{l: l, f: DebugString}
 }
 
 func (w *writer) Write(data []byte) (int, error) {
-	w.f(w.l, string(data))
-	return len(data), nil
+	var n int
+
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	i := bytes.IndexRune(data, '\n')
+	for i != -1 {
+		w.buf.Write(data[:i])
+		w.f(w.l, w.buf.String())
+		n += w.buf.Len()
+		w.buf.Reset()
+
+		data = data[i+1:]
```

Comment:
Unfortunately, there's no point in writing this data to the builder here, because the use of `Reset()` totally discards the internal buffer anyway.

```suggestion
		w.f(w.l, string(data[:i])
		data = data[i+1:]
```

Reply:
Oops, I just discarded the existing buffered content! I will amend my suggestions.